File: physics.js

function simulateTrajectory(params) {
    const g = 9.81; // gravity
    const dt = 0.1; // timestep

    let x = 0;
    let y = params.height || 0;

    let vx = params.speed * Math.cos(params.angle * Math.PI / 180);
    let vy = params.speed * Math.sin(params.angle * Math.PI / 180);

    const trajectory = [];
    let t = 0;

    while (y >= 0 && t < 300) {
        trajectory.push({ x, y, t });

        const speedTotal = Math.sqrt(vx * vx + vy * vy);
        const dragCoeff = 0.47;
        const area = Math.PI * (params.diameter / 2) ** 2;
        const drag = 0.5 * 1.225 * speedTotal * speedTotal * dragCoeff * area / params.mass;

        const ax = -drag * (vx / (speedTotal || 1));
        const ay = -g - drag * (vy / (speedTotal || 1));

        vx += ax * dt;
        vy += ay * dt;

        x += vx * dt;
        y += vy * dt;
        t += dt;
    }

    if (trajectory.length === 0) trajectory.push({ x: 0, y: params.height || 0, t: 0 });

    const xArr = trajectory.map(p => p.x);
    const yArr = trajectory.map(p => p.y);
    const tArr = trajectory.map(p => p.t);

    return {
        trajectory: { x: xArr, y: yArr, time: tArr },
        metrics: {
            max_range: Math.max(...xArr),
            max_height: Math.max(...yArr),
            impact_range: xArr[xArr.length - 1],
            total_flight_time: tArr[tArr.length - 1],
            impact_speed: Math.sqrt(vx * vx + vy * vy)
        }
    };
}

File: physics.py
import math

def simulate_trajectory(angle, speed, mass, diameter, height, wind_speed, wind_angle):
    g = 9.81
    dt = 0.1
    x = 0
    y = height
    vx = speed * math.cos(math.radians(angle))
    vy = speed * math.sin(math.radians(angle))

    trajectory = []
    t = 0

    while y >= 0 and t < 300:
        trajectory.append({'x': x, 'y': y, 't': t})
        speed_total = math.sqrt(vx**2 + vy**2)
        drag_coeff = 0.47
        area = math.pi * (diameter/2)**2
        drag = 0.5 * 1.225 * speed_total**2 * drag_coeff * area / mass
        ax = -drag * vx / speed_total
        ay = -g - drag * vy / speed_total
        vx += ax * dt
        vy += ay * dt
        x += vx * dt
        y += vy * dt
        t += dt

    metrics = {
        'max_range': max([p['x'] for p in trajectory]),
        'max_height': max([p['y'] for p in trajectory]),
        'impact_range': trajectory[-1]['x'],
        'total_flight_time': trajectory[-1]['t'],
        'impact_speed': math.sqrt(vx**2 + vy**2)
    }

    return {'trajectory': {'x':[p['x'] for p in trajectory], 'y':[p['y'] for p in trajectory], 'time':[p['t'] for p in trajectory]}, 'metrics': metrics}
